### PR TITLE
Overhaul, removing most semantic operations

### PR DESCRIPTION
**Pull request by patrickkwang**

### CODE REVIEW

File: strider/fetcher.py [hunk 1]
```diff
@@ -152,42 +143,81 @@ async def generate_plan(self):
...
+                        kp["url"],
+                        str(err),
+                    )
+                    self.kp_preferred_prefixes[kp["url"]] = dict()
+                    continue
+                try:
+                    meta_kg = response.json()
+                except JSONDecodeError as err:
+                    self.logger.warning(
+                        "Unable to get meta knowledge graph for KP %s: %s",
```
Comment (maximusunc) [inline strider/fetcher.py:179]: I would maybe make this a more descriptive error message. We were able to get the meta knowledge graph, but it came back malformed.

Comment (patrickkwang) [reply strider/fetcher.py]: Modified in 62ea25d

File: strider/query_planner.py [hunk 2]
```diff
@@ -431,138 +156,33 @@ async def generate_plans(
...
-
-    logger.info(
-        f"Found {num_duplicated_plans} plans, collapsed down to {len(plans)} plans")
-
-    if len(plans) == 0:
+    kps = dict()
+    for qedge_id in qgraph["edges"]:
+        qedge = qgraph["edges"][qedge_id]
+        provided_by = {"allowlist": None, "denylist": None} | qedge.get("provided_by", {})
+        kp_results = await kp_registry.search(
```
Comment (maximusunc) [inline strider/query_planner.py:168]: You could potentially use the `step_to_kps` function here.

File: strider/server.py [hunk 3]
```diff
@@ -159,11 +159,18 @@ def get_finished_query(
     results.expire(expiration_seconds)
     logs.expire(expiration_seconds)
 
+    # strip categories from node bindings
+    results = list(results.get())
+    for result in results:
+        for bindings in result["node_bindings"].values():
+            for binding in bindings:
+                binding.pop("category")
```
Comment (maximusunc) [inline strider/server.py:167]: Should this be `categories`?

File: strider/server.py [hunk 4]
```diff
@@ -412,21 +419,18 @@ async def extract_results(query_id, since, limit, offset, database):
...
+) -> list[list[str]]:
     """Generate plans for traversing knowledge providers."""
     query_graph = query.message.query_graph.dict()
 
-    await fill_categories_predicates(query_graph, logging.getLogger())
-    standardize_graph_lists(query_graph)
-    plans = await generate_plans(query_graph)
+    # await fill_categories_predicates(query_graph, logging.getLogger())
+    # standardize_graph_lists(query_graph)
+    plans, kps = await generate_plans(query_graph)
```
Comment (maximusunc) [inline strider/server.py:431]: `kps` is unused, and commented code could be removed.

File: strider/traversal.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
+"""Traversal."""
```
Comment (maximusunc) [inline strider/traversal.py:1]: This isn't very helpful.

File: strider/traversal.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        },
+        "edges": {
+            _qedge_id: qedge
+            for _qedge_id, qedge in qgraph["edges"].items()
+            if _qedge_id != qedge_id
+        }
+    }
+
+
+def get_traversals(qgraph: dict):
```
Comment (maximusunc) [inline strider/traversal.py:29]: Traversals can only include edges connected to pinned nodes? I would include that in the documentation if that's the case.

File: tests/test_query_planner.py [hunk 7]
```diff
@@ -489,30 +343,29 @@ async def test_unbound_unconnected_node(caplog):
 async def test_invalid_two_disconnected_components(caplog):
```
Comment (maximusunc) [inline tests/test_query_planner.py:343]: This name also should be changed to valid.

File: tests/test_server.py [hunk 8]
```diff
@@ -1571,15 +1516,15 @@ async def test_multiple_identifiers():
...
+        n0(( categories[] biolink:ChemicalSubstance ))
+        n1(( categories[] biolink:PhenotypicFeature ))
+        n1(( ids[] UMLS:C0032961 ))
         n0-- biolink:contraindicated_for -->n1
         """
     )
 
     # Create query
-    q = {"message" : {"query_graph" : QGRAPH}}
+    q = {"message" : {"query_graph" : QGRAPH}, "log_level": "DEBUG"}
```
Comment (maximusunc) [inline tests/test_server.py:1527]: Is the addition of the log level needed for these tests?

**Review by maximusunc — CHANGES_REQUESTED** [review]

File: strider/kp_registry.py [hunk 9]
```diff
@@ -71,26 +75,48 @@ async def delete_one(self, url):
...
         try:
             response = await post_json(
                 f'{self.url}/search',
                 {
-                    'source_type': source_types,
-                    'target_type': target_types,
-                    'edge_type': edge_types,
+                    'subject_category': subject_categories,
+                    'object_category': object_categories,
+                    'predicate': predicates,
```
[the remark this answers is not in the document]
Comment (patrickkwang) [reply strider/kp_registry.py:106]: No, the subject and object have to switch, too.

File: strider/fetcher.py [hunk 10]
```diff
@@ -152,42 +143,81 @@ async def generate_plan(self):
...
         })
 
         # Initialize results
 
-        first_step = next(iter(self.plan.keys()))
-        pinned_node_id = first_step.source
-        pinned_node = self.qgraph["nodes"][pinned_node_id]
+        first_qedge_id = self.plan[0]
+        # first_step = next(iter(self.plan.keys()))
+        # pinned_node_id = first_step.source
```
[the remark this answers is not in the document]
Comment (patrickkwang) [reply strider/fetcher.py]: d5743769a7f485a941ef414d79a999b4c7daa37d

File: strider/fetcher.py [hunk 11]
```diff
@@ -11,12 +11,16 @@
...
+from itertools import chain
 import json
+from json.decoder import JSONDecodeError
 import logging
 from datetime import datetime
 from typing import Optional
 
+import httpx
 from reasoner_pydantic import QueryGraph, Result, Response
+from reasoner_pydantic.results import NodeBinding
```
[the remark this answers is not in the document]
Comment (patrickkwang) [reply strider/fetcher.py]: ebe994610e05a5e5d00f02bc7c63affe63035b02

File: strider/query_planner.py [hunk 12]
```diff
@@ -7,9 +7,9 @@
 from reasoner_pydantic import QueryGraph
 
 from strider.kp_registry import Registry
-from strider.util import WrappedBMT
+from strider.util import WBMT
```
[the remark this answers is not in the document]
Comment (patrickkwang) [reply strider/query_planner.py]: 0933c0fcb0bc494b2a2227474d37a4243ea2814e

File: strider/query_planner.py (hunk 2 above)
Comment (patrickkwang) [reply strider/query_planner.py:167]: 023acde943e9a1a53674d5c4ec892482b5109de1

File: strider/server.py (hunk 3 above)
Comment (patrickkwang) [reply strider/server.py]: 938a893c352b0720e110031a0615fc7969eca083

File: strider/server.py (hunk 4 above)
Comment (patrickkwang) [reply strider/server.py]: 7c19566a15dd8ac747aacb39a870f9b02b7425c1

File: strider/traversal.py (hunk 5 above)
Comment (patrickkwang) [reply strider/traversal.py]: 77c3964a535622072698e8be3c438ec4ab04dae6

File: strider/traversal.py (hunk 6 above)
Comment (patrickkwang) [reply strider/traversal.py:29]: 77c3964a535622072698e8be3c438ec4ab04dae6

File: tests/test_query_planner.py (hunk 7 above)
Comment (patrickkwang) [reply tests/test_query_planner.py]: 77c3964a535622072698e8be3c438ec4ab04dae6

File: tests/test_server.py (hunk 8 above)
Comment (patrickkwang) [reply tests/test_server.py]: 636530af0af7ea75682cd320b717f67d4937988b

File: strider/fetcher.py [hunk 13]
```diff
@@ -200,46 +230,54 @@ def next_step(
...
-        # include descendants of any types that we received
-        all_valid_categories = []
-        for c in category:
-            all_valid_categories.extend(WBMT.get_descendants(c))
-
-        responses = await asyncio.gather(*(
-            self.portal.fetch(
-                kp["url"],
-                get_kp_request_body(
+        responses = await asyncio.gather(*chain(
```
[the remark this answers is not in the document]
Comment (patrickkwang) [reply strider/fetcher.py:247]: c42cb6bb8459e5ec42b9a8e0aefd8b4f745a0f49

File: strider/fetcher.py [hunk 14]
```diff
@@ -152,42 +143,81 @@ async def generate_plan(self):
...
 
         # Add partial results for each curie we are given
         for curie in curies:
             result = {
                 "node_bindings": {
-                    pinned_node_id: [{
+                    first_node_id: [{
                         "id": curie,
-                        "category": pinned_node["category"][0],
+                        "category": first_node["categories"][0],
```
[the remark this answers is not in the document]
Comment (patrickkwang) [reply strider/fetcher.py]: 8d706fe160f65b04e83f11400acac515dcbae2d6

File: strider/server.py [hunk 15]
```diff
@@ -161,10 +161,6 @@ def get_finished_query(
 
     # strip categories from node bindings
```
Comment (maximusunc) [inline strider/server.py:162]: This comment should be removed.

File: tests/test_server.py [hunk 16]
```diff
@@ -1421,7 +1363,10 @@ async def test_exception_response():
 
     response = await client.post(
         "/query",
-        json={"message": {"query_graph": qgraph}},
+        json={
+            "message": {"query_graph": qgraph},
+            "log_level": "DEBUG",
```
Comment (maximusunc) [inline tests/test_server.py:1368]: `log_level` can be removed in all tests.

File: strider/server.py (hunk 15 above)
Comment (patrickkwang) [reply strider/server.py]: fe0455dc3250e82c05c401530b1061ccbb8f70ac

File: tests/test_server.py (hunk 16 above)
Comment (patrickkwang) [reply tests/test_server.py:1369]: 64f0117eee0f1eaa6db46d8cec0461ac12788d6f

**Review by maximusunc — APPROVED** [review]
Looks good. I did try running test coverage, but pytest-cov isn't in the test dependencies.

**Pull request merged**